csi.proto

```diff
@@ -502,6 +556,213 @@ message ControllerServiceCapability {
+  // - Specify a policy to automatically clean up the snapshot.
+  // - Specify an expiration date for the snapshot.
+  // - Specify whether the snapshot is readonly or read/write.
+  // - Specify if the snapshot should be replicated to some place.
+  // - Specify primary or secondary for replication systems that
+  //   support snapshotting only on primary.
+  map<string, string> parameters = 5;
+
+  // NOTE: In the future, we may want to add “tags” to identify a
+  // snapshot. Purpose of tag is a "cookie" from CO and returned to CO.
+  // This is generated by CO internally and passed to the cloud storage
+  // backend through the plugin. This can be used to identify snapshots
+  // created by a specific CO.
+  // map<string, string> tags = 6;
+}
+
+message CreateSnapshotResponse {
+  // Contains all attributes of the newly created snapshot that are
+  // relevant to the CO along with information required by the Plugin
+  // to uniquely identify the snapshot. This field is REQUIRED.
+  Snapshot snapshot = 1;
+}
+
+// The information about a provisioned snapshot.
+message Snapshot {
+  // SnapshotSize contains `used_capacity_byes` which is the physical
+  // space consumed by the snapshot in bytes. `used_capacity_byes` may
+  // be variable over time. This field is OPTIONAL.
+  // If SnapshotSize is not set, it indicates this size is unknown.
+  SnapshotSize size = 1;
+
+  // Contains identity information for the created snapshot. This field
+  // is REQUIRED. The identity information will be used by the CO in
+  // subsequent calls to refer to the provisioned snapshot.
+  string id = 2;
+
+  // Identity information for the source volume. Note that create
+  // snapshot from snapshot is not supported here so the source has to
+  // be a volume. This field is REQUIRED.
+  string volume_id = 3;
+
+  // Timestamp when the point-in-time snapshot is taken on the storage
+  // system. This field is REQUIRED.
+  string timestamp = 4;
+
+  // The status of a snapshot. A snapshot could have the following
+  // statuses:
+  // CREATING - A snapshot is being cut.
+  // UPLOADING - A snapshot is cut and is now being uploaded.
+  // READY - A snapshot is ready for use.
+  // ERROR - Error occurred when creating or uploading the snapshot.
+  // Some cloud providers will upload the snapshot to a location in
+  // the cloud (i.e., an object store) after the snapshot is cut.
+  // Uploading is a long process that could take hours. At this phase,
+  // since the snapshot is already cut, a `thaw` operation can be
+  // performed so application can start running again (assuming a
+  // `freeze` operation was done before taking the snapshot). The
+  // status of the snapshot will become `READY` after the upload is
+  // complete.
+  // For cloud providers and storage systems that don't have the
+  // uploading process, the status should be `READY` after `CREATING`.
+  // `thaw` can be done when the status is `READY` in this case.
+  SnapshotStatus status = 5;
```

Comment:
How does user track that the snapshot is in creating state or stuck in creating? Is there something like a query snapshot state so that the CO can figure that either the snap completed or is stuck and issues a cancel op